src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,883 @@
+        this.#sendMessage(
+          {
+            action: IFrameMessageAction.LedgerSignTypedData,
+            params: {
+              hdPath,
+              domainSeparatorHex,
+              hashStructMessageHex,
+            },
+          },
+          (result) => resolve(result),
+        );
+      },
+    );
+
+    if (success && isSignMessageResponse(payload)) {
+      let v = parseInt(String(payload.v), 10).toString(16);
+      if (v.length < 2) {
+        v = `0${v}`;
+      }
+      const signature = `0x${payload.r}${payload.s}${v}`;
+      // @types/eth-sig-util seems to be wrong
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore next-line
+      const addressSignedWith = sigUtil.recoverTypedSignature_v4({
+        data,
+        sig: signature,
+      });
+      if (
+        ethUtil.toChecksumAddress(addressSignedWith) !==
+        ethUtil.toChecksumAddress(withAccount)
+      ) {
+        throw new Error('Ledger: The signature doesnt match the right address');
+      }
+      return signature;
+    }
+    throw (
+      payload.error || new Error('Ledger: Unknown error while signing message')
+    );
+  }
+
+  exportAccount() {
+    throw new Error('Not supported on this device');
+  }
+
+  forgetDevice() {
+    this.accounts = [];
+    this.page = 0;
+    this.unlockedAccount = 0;
+    this.paths = {};
+    this.accountDetails = {};
+    this.hdk = new HDKey();
+  }
+
+  /* PRIVATE METHODS */
+
+  #setupIframe() {
+    this.iframe = document.createElement('iframe');
+    this.iframe.src = this.bridgeUrl;
+    this.iframe.allow = `hid 'src'`;
+    this.iframe.onload = async () => {
+      // If the ledger live preference was set before the iframe is loaded,
+      // set it after the iframe has loaded
+      this.iframeLoaded = true;
+      if (this.delayedPromise) {
+        try {
+          const result = await this.updateTransportMethod(
+            this.delayedPromise.transportType,
+          );
+          this.delayedPromise.resolve(result);
+        } catch (e) {
+          this.delayedPromise.reject(e);
+        } finally {
+          delete this.delayedPromise;
+        }
+      }
+    };
+    document.head.appendChild(this.iframe);
+  }
+
+  #getOrigin() {
+    const tmp = this.bridgeUrl.split('/');
+    tmp.splice(-1, 1);
+    return tmp.join('/');
+  }
+
+  #sendMessage(
+    msg: Partial<IFrameMessage>,
+    cb: (response: IFrameMessageResponse) => void,
+  ) {
+    msg.target = 'LEDGER-IFRAME';
+
+    this.currentMessageId += 1;
+    msg.messageId = this.currentMessageId;
+
+    this.messageCallbacks[this.currentMessageId] = cb;
+    this.iframe?.contentWindow?.postMessage(msg, '*');
+  }
+
+  #setupListener() {
+    this.#eventListener = ({ origin, data }) => {
+      if (origin !== this.#getOrigin()) {
+        return false;
+      }
+
+      if (data) {
+        if (this.messageCallbacks[data.messageId]) {
+          this.messageCallbacks[data.messageId]?.(data);
```

Comment:
You shouldn't need the `?.` here since you just checked it. If TypeScript is still complaining this may be undefined, I believe you can get around this by first assigning `this.messageCallbacks[data.messageId]` to a variable.